rafthttp/pipeline.go

```diff
@@ -128,31 +132,43 @@ func (p *pipeline) post(data []byte) (err error) {
 		}
 	}()
 
-	resp, err := p.tr.pipelineRt.RoundTrip(req)
-	done <- struct{}{}
-	if err != nil {
-		p.picker.unreachable(u)
-		return err
-	}
-	b, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		p.picker.unreachable(u)
-		return err
-	}
-	resp.Body.Close()
-
-	err = checkPostResponse(resp, b, req, p.peerID)
-	if err != nil {
-		p.picker.unreachable(u)
-		// errMemberRemoved is a critical error since a removed member should
-		// always be stopped. So we use reportCriticalError to report it to errorc.
-		if err == errMemberRemoved {
-			reportCriticalError(err, p.errorc)
+	go func() {
```

Comment:
I do not think I understand what you said.

Let me clarify it.

now, we:

- create context with cancel at line 122
- pass in the created context to req at line 123

Now if the cancel fun get called, the request should be canceled.

- create a go routine to listen on stop. if server is stopped, we can cancel.

So if server is stopped, go routine should be triggered on stop channel, and call cancel.

After it calls cancel, the http request should be canceled.

From your test, it fails to cancel the http request. There is something goes wrong.

You "fix" that by making the http request nonblocking. That is not what we want. We want the request to be canceled once the server is stopped. Basically we need to make sure the go routine created at 134 exits.